### ISSUE TITLE
Font size calculations inacurate

### ISSUE THREAD

**Issue by johanvdb** [openfl/openfl#908]
Hi

The following scenario shows inconsistent results:
1. Create an empty test app
2. Load all the fonts via Assets.list (use a few fonts from google fonts as test)
3. Create a TextField for each font and showcase the font's tallest characters
4. Draw a rectangle around each TextField making use of the "textHeight" and "textWidth" properties (make it 2px bigger on all sides).

Source code is available at https://github.com/johanvdb/openfl-test-fonts

I realise that this is partially being discussed under #288, but I believe that this is wider than just html5, and while the biggest problems are currently being experienced in html5, I hope to see all platforms behave the same at the end of this exercise.

Here are my findings:

```
openfl test mac -Dlegacy
```

MacBook Pro with OSX 10.9.5
<img width="895" alt="screen shot 2015-10-28 at 1 49 01 pm" src="https://cloud.githubusercontent.com/assets/815840/10787592/bf3c075a-7d7a-11e5-82e0-1070867836f1.png">

```
openfl test mac 
```

MacBook Pro with OSX 10.9.5
<img width="800" alt="screen shot 2015-10-28 at 1 56 43 pm" src="https://cloud.githubusercontent.com/assets/815840/10787755/d4781716-7d7b-11e5-9465-535e536c00c1.png">

```
openfl test html5
```

Chrome Version 46.0.2490.80 (64-bit)
<img width="851" alt="screen shot 2015-10-28 at 1 59 24 pm" src="https://cloud.githubusercontent.com/assets/815840/10787837/38d354c8-7d7c-11e5-878f-4544a496691b.png">

```
openfl test html5
```

Safari 7.1.6
<img width="685" alt="screen shot 2015-10-28 at 2 00 36 pm" src="https://cloud.githubusercontent.com/assets/815840/10787855/51dcf960-7d7c-11e5-8ef1-bfac93565a22.png">

```
openfl test html5 -Ddom
```

Chrome Version 46.0.2490.80 (64-bit)
<img width="839" alt="screen shot 2015-10-28 at 2 05 58 pm" src="https://cloud.githubusercontent.com/assets/815840/10787999/1f658a82-7d7d-11e5-8b09-b3b8aa6688b5.png">

```
openfl test html5 -Dwebgl
```

Chrome Version 46.0.2490.80 (64-bit)
<img width="1118" alt="screen shot 2015-10-28 at 2 07 26 pm" src="https://cloud.githubusercontent.com/assets/815840/10788018/4adad4e2-7d7d-11e5-8fd8-c0526672de13.png">

```
openfl test neko
```

<img width="798" alt="screen shot 2015-10-28 at 2 08 30 pm" src="https://cloud.githubusercontent.com/assets/815840/10788040/711b0c8a-7d7d-11e5-85e9-a6c73c781aa0.png">

```
johantheitguy$ haxelib list
actuate: [1.8.6]
box2d: [1.2.3]
flixel-addons: [1.1.0]
flixel-demos: [1.1.2]
flixel-tools: 1.1.0 [1.1.1]
flixel: [3.3.11] 3.3.8
hxargs: [3.0.0]
hxcpp: 3.2.102 [3.2.180]
layout: 1.2.0 [1.2.1]
lime-samples: 2.3.0 [2.6.0]
lime: 2.4.3 2.4.4 2.4.5 2.6.0 2.6.1 2.6.2 [2.6.8]
nape: [2.0.19]
openfl-bitfive: 3.0.0 3.3.2 [3.3.3]
openfl-samples: 2.2.2 3.1.0 3.3.0 [3.3.1]
openfl: 3.0.1 3.1.0 3.3.1 3.3.2 3.3.3 [3.3.8]
swf: 1.8.8 1.9.1 2.0.2 [2.1.1]
```

**Comment by johanvdb** [comment]
As you can see, there is something terribly wrong with the handling of "RockSalt" on html5 canvas. On most other platforms, the leading space for most fonts are differing, and it would seem as if Neko is the only one that handles it correctly. I forgot to include a flash build:

<img width="569" alt="screen shot 2015-10-28 at 2 20 53 pm" src="https://cloud.githubusercontent.com/assets/815840/10788318/4d740cb2-7d7f-11e5-9ea7-9edf4fd569ab.png">

Flash also seems fairly ok, although all platforms consistently do not cater for the "g" and "j" correctly and overflows by 2 pixels at this size for the OpenSans, Press Start and default fonts.

**Comment by johanvdb** [comment]
I will try to implement https://github.com/nodebox/opentype.js as a test case into openfl.

At present, when the target platform has no way of calculating the height, openfl's TextEngine trusts the "size" parameter of the TextFormat and calculates a height off that. https://github.com/openfl/openfl/blob/3.4.0/openfl/_internal/text/TextEngine.hx#L693-L743

Implementing opentype.js would provide us with the needed "font.unitsPerEM" functionality to calculate the correct size.

**Comment by ShaheedLegion** [comment]
Bump.

**Comment by chatziko** [comment]
Motivated by #1801, I made some changes to include font metrics in html5 for fonts known at build time, and also to allow manually setting the metrics for unknown fonts (more info in #1801). I also changed `TextEngine` to use the metrics in html5 if available, which seems to fix this issue.

I tried the app reported here, I get almost identical results in all html5-firefox / html5-chrome / flash / linux / neko ([screenshots here](https://github.com/openfl/openfl/files/1643884/openfl-test-fonts.zip)).

The branches are [here](https://github.com/openfl/openfl/compare/develop...chatziko:html5-font-metrics) and [here](https://github.com/openfl/lime/compare/develop...chatziko:html5-font-metrics) for anyone who wants to test it.

**Comment by chatziko** [comment]
Just wanted to point out that, although this issue was closed, the problem is **not** fixed unless #1814 is also merged.

Currently (v7.1.2) font metrics are included in html5 (PR openfl/lime#1152) but they're not used anywhere!
